app/src/main/java/org/thoughtcrime/securesms/pin/KbsRepository.java

```
@@ -0,0 +1,141 @@
+                                                                   @NonNull TokenResponse tokenResponse)
+      throws IOException, KeyBackupSystemWrongPinException, KeyBackupSystemNoDataException
+  {
+    Log.i(TAG, "restoreMasterKey()");
+
+    if (pin == null) return null;
+
+    if (basicStorageCredentials == null) {
+      throw new AssertionError("Cannot restore KBS key, no storage credentials supplied");
+    }
+
+    Log.i(TAG, "Preparing to restore from " + enclave.getEnclaveName());
+    return restoreMasterKeyFromEnclave(enclave, pin, basicStorageCredentials, tokenResponse);
+  }
+
+  private static @NonNull KbsPinData restoreMasterKeyFromEnclave(@NonNull KbsEnclave enclave,
+                                                                 @NonNull String pin,
+                                                                 @NonNull String basicStorageCredentials,
+                                                                 @NonNull TokenResponse tokenResponse)
+      throws IOException, KeyBackupSystemWrongPinException, KeyBackupSystemNoDataException
+  {
+    KeyBackupService                keyBackupService = ApplicationDependencies.getKeyBackupService(enclave);
+    KeyBackupService.RestoreSession session          = keyBackupService.newRegistrationSession(basicStorageCredentials, tokenResponse);
+
+    try {
+      Log.i(TAG, "Restoring pin from KBS");
+
+      HashedPin  hashedPin = PinHashing.hashPin(pin, session);
+      KbsPinData kbsData   = session.restorePin(hashedPin);
+
+      if (kbsData != null) {
+        Log.i(TAG, "Found registration lock token on KBS.");
+      } else {
+        throw new AssertionError("Null not expected");
+      }
+
+      return kbsData;
+    } catch (UnauthenticatedResponseException | InvalidKeyException e) {
+      Log.w(TAG, "Failed to restore key", e);
+      throw new IOException(e);
+    } catch (KeyBackupServicePinException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)